imports/client/components/FancyEditor.tsx

```diff
@@ -0,0 +1,639 @@
+// Normal editing can't create more than one, but copy-pasting a break would
+// normally get transformed into a hard break, so we solve this by merging
+// sibling message nodes.
+const withSingleMessage = (editor: Editor) => {
+  const { normalizeNode } = editor;
+  editor.normalizeNode = (entry: NodeEntry) => {
+    const [node, _path] = entry;
+    if (Editor.isEditor(node)) {
+      if (node.children.length > 1) {
+        Transforms.mergeNodes(editor, { at: [1] });
+      }
+    }
+
+    normalizeNode(entry);
+  };
+  return editor;
+};
+/* eslint-enable no-param-reassign */
+
+const insertMention = (editor: Editor, userId: string) => {
+  const mention: MentionElement = {
+    type: 'mention',
+    userId,
+    children: [{ text: '' }],
+  };
+  Transforms.insertNodes(editor, mention);
+  Transforms.move(editor);
+};
+
+const MatchCandidate = ({
+  user, selected, onSelected,
+}: {
+  user: Meteor.User,
+  selected: boolean,
+  onSelected: (u: Meteor.User) => void,
+}) => {
+  const onClick = useCallback(() => {
+    onSelected(user);
+  }, [onSelected, user]);
+  return (
+    <div
+      key={user._id}
+      style={{
+        padding: '1px 3px',
+        borderRadius: '3px',
+        background: selected ? '#e0ecfc' : 'transparent',
+      }}
+      onClick={onClick}
+    >
+      {user.displayName}
+    </div>
+  );
+};
+
+type AugmentedUser = Meteor.User & {
+  foundDisplayName: boolean;
+  startsDisplayName: boolean;
+  foundEmail: boolean;
+  startsEmail: boolean;
+  foundGoogle: boolean;
+  startsGoogle: boolean;
+  foundDiscord: boolean;
+  startsDiscord: boolean;
+};
+
+function matchUsers(users: Meteor.User[], search: string): AugmentedUser[] {
+  // No point doing all this matching work if there's no search string to match against.
+  if (!search) return [];
+
+  // We approximate case-insensitive search by converting both needles and
+  // haystacks to lowercase.
+  const needle = search.toLowerCase();
+
+  // Augment users with whether assorted fields match the search string or not, and where.
+  const augmentedUsers: AugmentedUser[] = users.map((u) => {
+    // Note which user fields case-insensitively match the current search
+    // string, and whether that match was at the start of the field.
+    const displayName = u.displayName?.toLowerCase() ?? '';
+    const foundDisplayName = displayName.includes(needle);
+    const startsDisplayName = displayName.startsWith(needle);
+    const emails = u.emails?.map((addrObj) => addrObj.address.toLowerCase()) ?? [];
+    const foundEmail = emails.some((addr) => addr.includes(needle));
+    const startsEmail = emails.some((addr) => addr.startsWith(needle));
+    const googEmail = u.googleAccount?.toLowerCase() ?? '';
+    const foundGoogle = googEmail.includes(needle);
+    const startsGoogle = googEmail.startsWith(needle);
+    const discordName = u.discordAccount ? `${u.discordAccount.username.toLowerCase()}#${u.discordAccount.discriminator}` : '';
```

Comment:
Just as a note I opened #1312 as vaguely related to this, though not worth blocking over

Reply:
Yeah, that seems both reasonable and largely orthogonal to this work.